### ISSUE TITLE
Fix Differences in Generated Clients

### ISSUE THREAD

**Issue by noheton** [noheton/shepard#136]
***In GitLab by @xit-max-hey on Aug 7, 2024, 11:04 GMT+2:***

There are some issues that need to be resolved:

* The openapi-diff job is failing right now complaining about property types: https://gitlab.com/dlr-shepard/shepard/-/jobs/7524100176undefined
* Although the behaviour of the endpoints and the data types are the same with Quarkus, the generated OpenAPI spec has some differences in it's schema partitioning. That results in different models being generated durign client generation, which implies breaking changes for users of the generated clients. This can be reproduced by using version **2024.8.8-dev.1530** of the shepard client in the frontend.

**Assignees:** xit-max-hey

*Migrated from GitLab: https://gitlab.com/dlr-shepard/shepard/-/work_items/136*

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Aug 7, 2024, 11:04 GMT+2:***

added #73 as parent issue

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Aug 8, 2024, 14:15 GMT+2:***

added #84 as parent issue

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Aug 9, 2024, 10:05 GMT+2:***

Possible Solutions:

<table>
<tr>
<th>Solution</th>
<th>Notes</th>
<th>Feasibility</th>
</tr>
<tr>
<td>Dereference schemas with yaml dereferencing</td>
<td>

* using some tool
* the OpenAPI diff job cannot resolve the references and fails
* Introduces new issues for client generation because request types are lost
</td>
<td>No</td>
</tr>
<tr>
<td>Dereference schemas with yaml dereferencing before client generation</td>
<td>

* Clients may work and produce the correct models
* The normal OpenAPI should stay unchanged and the OpenAPi-diff job should work
* introduces new issues for client generation because request types are lost
</td>
<td>No</td>
</tr>
<tr>
<td>Manually dereference schemas</td>
<td>

* Manually adjust references so that the generated clients stay the same
* A script would need to be written
* we may miss some relevant parts
</td>
<td>Maybe, but a lot of effort for a workaround and risk of missing parts</td>
</tr>
<tr>
<td>Adjust Schema names in the code</td>
<td>

* Replace names of Schema annotations in the code
* we may miss some relevant parts
</td>
<td>Maybe, but a lot of effort for a workaround and risk of missing parts</td>
</tr>
<tr>
<td>

Use https://quarkus.io/guides/openapi-swaggerui#enhancing-the-openapi-schema-with-filters to adapt schema
</td>
<td>

* we may miss some relevant parts
</td>
<td>Maybe, but a lot of effort for a workaround and risk of missing parts</td>
</tr>
<tr>
<td>Ship old OpenAPI spec</td>
<td>

* Low effort
* Generated clients stay the same
* Hacky solution
* future developments would imply manual adaptions
</td>
<td>Low effort, may be feasible if it's only a workaround for a few releases</td>
</tr>
<tr>
<td>Generate clients with old OpenAPI spec</td>
<td>

* Low effort
* Generated clients stay the same
* Hacky solution
* OpenAPI spec and Swagger-UI of backend will change -\> May lead to changes for consumers interacting with the openapi spec directly
</td>
<td>Low effort, may be feasible if it's only a workaround for a few releases</td>
</tr>
<tr>
<td>Don't publish new clients</td>
<td>

* Old clients do still work unless there are breaking API changes
* Users of the new versioning feature would need to use a newer "experimental" client with breaking changes
</td>
<td>Low effort, may be feasible if it's only a workaround for a few releases</td>
</tr>
<tr>
<td>Introduce breaking change</td>
<td>

</td>
<td>Users may not be satisfied</td>
</tr>
<tr>
<td>Switch from MicroProfile back to Swagger annotations</td>
<td>

* Basically reverting the changes contained in this commit: https://gitlab.com/dlr-shepard/shepard/-/commit/7dfc1393c92c74990e10f73d984b399c70decb67 and some more changes
* not sure if this works without issues
</td>
<td>Not the recommended way for openapi annotations in quarkus</td>
</tr>
</table>

Maybe we can discuss how to go forward on monday, @thaase @xit-max-hey

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Aug 9, 2024, 11:25 GMT+2:***

To document what is the issue here, I'm going to explain one issue here:
Below is an image comparing the two generated openapi specs from the quarkus backend (left) and the old backend (right).
undefined

We see that old backend defines a "Permissions" object with multiple properties. One property is the permiussion type, which is a string enum with three values.
Quarkus generates almost exactly the same openapi spec, but instead of defining the enum inline, it generates the "permissionType" as a reference to the "PermissionType" object, which represents the same 3-value string enum.
Function wise these specs should be equal, since the $ref is replaced by the "PermissionType" object.

However, if we take a look at the generated client code, some important differences are visisble:

undefined

In the above image the left side is the client generated by the Quarkus openapi spec, and on the right side is the client code generated by the old backend openapi spec.

The client code is almost identical, except for the naming.
The client code on the right side defines the "permissionType" as a "PermissionsPermissionTypeEnum", which is defined in the same file.
However, the client code on the left side uses as the "permissionType" a "PermissionType" object, which is defined in another file. This object is only named "PermissionType".

**There, breaking change is introduced**, since if the frontend code wants to use the Permissions interface, it currently expects a "PermissionsPermissionTypeEnum", however the enum generated by the quarkus openapi spec provides only a "PermissionType" enum. So the names of these enum types differ, which is a breaking change.

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Aug 12, 2024, 09:52 GMT+2:***

Decision 12.08: We go with the breaking change and make sure to properly test & document it before the september release. We created a new issue for it #143

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Aug 12, 2024, 09:53 GMT+2:***

marked this task as related to #143
